### PR TITLE
Fix error when changing route from virtual appliance to any other type

### LINKED ISSUE #1352: Cannot change route from VirtualAppliance to None
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.azurerm v1.6.0
```

### Affected Resource(s)

* azurerm_route_table

### Terraform Configuration Files

Before
```hcl
resource "azurerm_route_table" "primary" {
  name                = "primary"
  location            = "${azurerm_resource_group.routes.location}"
  resource_group_name = "${azurerm_resource_group.routes.name}"

  route {
    name                   = "UDR_to_InternetProxy"
    address_prefix         = "0.0.0.0/0"
    next_hop_type          = "VirtualAppliance"
    next_hop_in_ip_address = "192.168.0.1"
  }
}
```

After
```hcl
resource "azurerm_route_table" "primary" {
  name                = "primary"
  location            = "${azurerm_resource_group.routes.location}"
  resource_group_name = "${azurerm_resource_group.routes.name}"

  route {
    name           = "UDR_BlackHole"
    address_prefix = "0.0.0.0/0"
    next_hop_type  = "None"
  }
}
```

### Expected Behavior

Route is updated

### Actual Behavior

```
1 error(s) occurred:

* azurerm_route_table.primary: 1 error(s) occurred:

* azurerm_route_table.primary: Error Creating/Updating Route Table "primary" (Resource Group "Network-Routes"): network.RouteTablesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="NextHopIpAddressNotAllowed" Message="NextHopIpAddress 192.168.0.1 for route UDR_BlackHole cannot be specified. This field can only be specified for NextHopType VirtualAppliance." Details=[]
```

### Steps to Reproduce

1. `terraform apply` the "before" config
2. `terraform apply` the "after" config

### PR DESCRIPTION
**Pull request by draggeta**
There was a problem with changing from virtual appliance to any other type. This was caused by the fact that `next_hop_in_ip_address` was computed.

The issue occurred in both azurerm_route_table and azurerm_route.

Also made some other changes to make both resources similar.

- [x] fix resource
- [x] add tests to check for regression in future

Closes #1352

### CODE REVIEW

**Comment by draggeta** [comment]
There are a few linting errors from other resources. Those are optimizations. Should I fix them here or should they be fixed in separate pull requests?

**Comment by tombuildsstuff** [comment]
@draggeta sorry about the linting errors, they've been fixed via #2190 - I've re-run the build now that's been merged and this now passes 👍

**Comment by draggeta** [comment]
@tombuildsstuff No worries, I was the cause of one of them anyway 😊

File: azurerm/resource_arm_route_test.go [hunk 1]
```diff
@@ -35,6 +35,41 @@ func TestAccAzureRMRoute_basic(t *testing.T) {
...
+					testCheckAzureRMRouteExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "next_hop_type", "VirtualAppliance"),
+					resource.TestCheckResourceAttr(resourceName, "next_hop_in_ip_address", "192.168.0.1"),
+				),
+			},
+			{
+				Config: testAccAzureRMRoute_basic(ri, testLocation()),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMRouteExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "next_hop_type", "VnetLocal"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_route_test.go]: we should also be able to add in:
```suggestion
					resource.TestCheckResourceAttr(resourceName, "next_hop_type", "VnetLocal"),
					resource.TestCheckResourceAttr(resourceName, "next_hop_in_ip_address", ""),
```

to ensure this field isn't set

File: azurerm/resource_arm_route_test.go [hunk 2]
```diff
@@ -35,6 +35,41 @@ func TestAccAzureRMRoute_basic(t *testing.T) {
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMRouteDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAzureRMRoute_basic(ri, testLocation()),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMRouteExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "next_hop_type", "VnetLocal"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_route_test.go]: we should also be able to add in:

```suggestion
					resource.TestCheckResourceAttr(resourceName, "next_hop_type", "VnetLocal"),
					resource.TestCheckResourceAttr(resourceName, "next_hop_in_ip_address", ""),
```

to ensure this field isn't set

File: azurerm/resource_arm_route_test.go (hunk 1 above)
Comment (draggeta) [reply azurerm/resource_arm_route_test.go]: Good idea. I was being implicit, while being explicit here might be better.

File: azurerm/resource_arm_route_test.go (hunk 2 above)
Comment (draggeta) [reply azurerm/resource_arm_route_test.go]: Done

**Comment by tombuildsstuff** [comment]
hey @draggeta 

Running the tests I noticed one of them is failing due to the `next_hop_in_ip_address` field being conditionally set (and so this isn't set to an empty string):

```
------- Stdout: -------
=== RUN   TestAccAzureRMRoute_update
--- FAIL: TestAccAzureRMRoute_update (76.21s)
	testing.go:538: Step 0 error: Check failed: Check 3/3 error: azurerm_route.test: Attribute 'next_hop_in_ip_address' not found
FAIL
```

So that we can get this merged I've pushed a commit to fix this - I hope you don't mind!

Thanks!

**Comment by tombuildsstuff** [comment]
Tests pass:

screenshot 2018-11-07 at 13 29 19

**Comment by moisedo** [comment]
So in version 0.18 it was allowing to have this next_hop_in_ip_address null, but due to the verification it now needs a value if its specified. The big issue with this is, when you have N amount of routes per route table using a MAP, you cannot use a TF resource template anymore, you will have to create 2 templates (one for next_hop_in_ip_address and one without it) Is there another way to tackle this without duplicating code? Thank you. It will also allow me to understand the reason behind mandating a value on this key when is specified.

**Pull request merged**